target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c

```diff
@@ -136,14 +136,6 @@ static void rtl83xx_vlan_setup(struct rtl838x_switch_priv *priv)
 	for (int i = 0; i < MAX_VLANS; i ++)
 		priv->r->vlan_set_tagged(i, &info);
 
-	/* reset PVIDs; defaults to 1 on reset */
-	for (int i = 0; i <= priv->ds->num_ports; i++) {
-		priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_INNER, 0);
-		priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_OUTER, 0);
-		priv->r->vlan_port_pvidmode_set(i, PBVLAN_TYPE_INNER, PBVLAN_MODE_UNTAG_AND_PRITAG);
-		priv->r->vlan_port_pvidmode_set(i, PBVLAN_TYPE_OUTER, PBVLAN_MODE_UNTAG_AND_PRITAG);
-	}
```

Comment:
Yes I think the driver is setting pvid to 0
OpenWRT shows PVID 1 with `bridge vlan` command


And considering rtlsdk, they also set pvid=1 by default, 
https://gitlab.com/olliver/openwrt/realtek_sdk/-/blob/xgs1250/sdk/include/rtk/default.h?ref_type=heads#L46

Reply:
Some more investigation before this commit,
on the first boot, following network config is generated,
```
config interface 'loopback'                                                                                                                                                                                                                                   
        ...                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                              
config globals 'globals'                                                                                                                                                                                                                                      
        ...                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                              
config device 'switch'                                                                                                                                                                                                                                        
        option name 'switch'                                                                                                                                                                                                                                  
        option type 'bridge'                                                                                                                                                                                                                                  
        option macaddr '...'                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
config bridge-vlan 'lan_vlan'                                                                                                                                                                                                                                 
        option device 'switch'                                                                                                                                                                                                                                
        option vlan '1'                                                                                                                                                                                                                                       
        option ports 'lan0 lan1 lan2 lan3 lan4 lan5 lan6 lan7'                                                                                                                                                                                                
                                                                                                                                                                                                                                                              
config device                                                                                                                                                                                                                                                 
        option name 'switch.1'                                                                                                                                                                                                                                
        option macaddr '...'                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
config interface 'lan'                                                                                                                                                                                                                                        
        option device 'switch.1'                                                                                                                                                                                                                              
        option proto 'static'                                                                                                                                                                                                                                 
        option ipaddr '192.168.1.1'                                                                                                                                                                                                                           
        option netmask '255.255.255.0'                                                                                                                                                                                                                        
        option ip6assign '60'                                                                                                                                                                                                                                 
                                                    
```
In this configuration `bridge vlan` shows following output,
```
root@OpenWrt:/# bridge vlan 
port              vlan-id  
lan0              1 PVID Egress Untagged
lan1              1 PVID Egress Untagged
lan2              1 PVID Egress Untagged
lan3              1 PVID Egress Untagged
lan4              1 PVID Egress Untagged
lan5              1 PVID Egress Untagged
lan6              1 PVID Egress Untagged
lan7              1 PVID Egress Untagged
switch            1
```
And in this state, I am not able to ping a device connected to one of the lan ports from switch, `ping 192.168.1.100` no packets are received on any of the lan ports

---

If I make following changes to the `/etc/config/network`,
```
--- a/etc/config/network      2024-01-03 13:30:20.130108139 +0100
+++ b/etc/config/network      2024-01-03 13:30:36.321918906 +0100
@@ -10,19 +10,11 @@
 config device 'switch'
         option name 'switch'
         option type 'bridge'
-        option macaddr '...'
-
-config bridge-vlan 'lan_vlan'
-        option device 'switch'
-        option vlan '1'
         option ports 'lan0 lan1 lan2 lan3 lan4 lan5 lan6 lan7'
-
-config device
-        option name 'switch.1'
         option macaddr '...'
 
 config interface 'lan'
-        option device 'switch.1'
+        option device 'switch'
         option proto 'static'
         option ipaddr '192.168.1.1'
         option netmask '255.255.255.0'

```
output of `bridge vlan`,
```
root@OpenWrt:/# bridge vlan                                                                                                                                                                                                                                   
port              vlan-id                                                                                                                                                                                                                                     
lan0              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan1              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan2              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan3              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan4              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan5              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan6              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan7              1 PVID Egress Untagged                                                                                                                                                                                                                      
switch            1 PVID Egress Untagged          
```
Notice that switch vlan-id 1 is now PVID Egress Untagged.
After this, I am able to ping the device connected to one of the lan ports.

---
If I remove reset-pvid for-loop I am able to ping the connected device just fine without making any changes. and here I am not sure if the default config is problematic or the driver code. 
Also it looks like driver is setting pvid 0, but default config comes up with pvid 1.

Reply:
Well I just spotted the for-loop goes over 28 ports, and that was the issue.
- reset pvid is anyways overwritten by the configuration so this is misinterpretation on my side of setup code.
```
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
index 9b26231a71..83bf92c8d7 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
@@ -137,7 +137,7 @@ static void rtl83xx_vlan_setup(struct rtl838x_switch_priv *priv)
                priv->r->vlan_set_tagged(i, &info);
 
        /* reset PVIDs; defaults to 1 on reset */
-       for (int i = 0; i <= priv->ds->num_ports; i++) {
+       for (int i = 0; i < priv->cpu_port; i++) {
                priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_INNER, 0);
                priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_OUTER, 0);
                priv->r->vlan_port_pvidmode_set(i, PBVLAN_TYPE_INNER, PBVLAN_MODE_UNTAG_AND_PRITAG);
```

before the code goes up to i = 29, which calculates the register value to `0x834c` which is wrong, but still not the root cause since that register is being overwritten by vlan_fwd_on_inner.
Problems is when register 0x8348 (VLAN_PORT_PB_VLAN[28]) is set with 0x000000 with
```
priv->r->vlan_port_pvid_set(i:28, PBVLAN_TYPE_INNER, 0);
priv->r->vlan_port_pvid_set(i:28, PBVLAN_TYPE_OUTER, 0);
priv->r->vlan_port_pvidmode_set(i:28, PBVLAN_TYPE_INNER, PBVLAN_MODE_UNTAG_AND_PRITAG);
priv->r->vlan_port_pvidmode_set(i:28, PBVLAN_TYPE_OUTER, PBVLAN_MODE_UNTAG_AND_PRITAG);
```
If I manually set register `0x834c` to `0x00010004` (OPVID:1,IPVID:1) ping starts working. bootloader and rtlsdk sets (OPVID:1, IPVID:1) so what do you suggest the code should do? set default pvid to 1 for all ports or do this just for cpu port?

Reply:
use reset pvid 1 instead of 0, check latest commit's message and above investigation for explanation.